src/runtime_src/core/common/usage_metrics.cpp

```diff
@@ -0,0 +1,525 @@
+}
+
+// Create specific logger if ini option is enabled
+static std::shared_ptr<base_logger>
+get_logger_object()
+{
+  if (xrt_core::config::get_usage_metrics_logging())
+    return std::make_shared<usage_metrics_logger>();
+
+  return std::make_shared<base_logger>();
+}
+
+// Per thread logger object  
+std::shared_ptr<base_logger>
+get_usage_metrics_logger()
+{
+  static thread_local auto usage_logger_object = get_logger_object();
+  return usage_logger_object;
+}
+
+usage_metrics_logger::
+usage_metrics_logger() : map_ptr(usage_metrics_map)
+{
+  thread_count++;
+}
+
+usage_metrics_logger::
+~usage_metrics_logger()
+{
+  {
+    std::lock_guard<std::mutex> lk(m);
+    thread_count--;
+
+    // push this threads usage metrics to global map
+    // in thread safe manner
+    (*map_ptr)[std::this_thread::get_id()] = std::move(m_dev_map);
+  }
+
+  if (thread_count == 0) {
+    // print usage metrics log after all threads are destroyed
+    print_usage_metrics();
+  }
+}
+
+void
+usage_metrics_logger::
+log_device_info(const xrt_core::device* dev)
+{
+  if (auto dev_id = dev->get_device_id(); m_dev_map.find(dev_id) == m_dev_map.end()) {
+    m_dev_map[dev_id] = {};
+    try {
+      auto bdf = xrt_core::query::pcie_bdf::to_string(xrt_core::device_query<xrt_core::query::pcie_bdf>(dev));
+      m_dev_map[dev_id].bdf = bdf;
+    }   
+    catch (...) {}
+  }
+}
+
+void 
+usage_metrics_logger::
+log_hw_ctx_info(const xrt::hw_context_impl* hwctx_impl)
+{
+  try {
+    auto hw_ctx = 
+        xrt_core::hw_context_int::create_hw_context_from_implementation(const_cast<xrt::hw_context_impl*>(hwctx_impl));
+    
+    auto hwctx_handle = static_cast<xrt_core::hwctx_handle*>(hw_ctx);
+    auto dev_id = xrt_core::hw_context_int::get_core_device(hw_ctx)->get_device_id();
+    auto uuid = hw_ctx.get_xclbin_uuid();
+
+    auto map_it = m_dev_map.find(dev_id);
+    // dont log if device didn't match
+    if (map_it == m_dev_map.end())
+      return;
+    
+    auto& [device_id, dev_metrics] = *map_it;
+    auto ctx_it = std::find_if(dev_metrics.hw_ctx_vec.begin(), dev_metrics.hw_ctx_vec.end(), 
+                              [hwctx_handle](hw_ctx_metrics& ctx) 
+                              {
+                                return ctx.handle == hwctx_handle;
+                              }
+                              );
+    
+    if (ctx_it == dev_metrics.hw_ctx_vec.end())
+      dev_metrics.hw_ctx_vec.emplace_back(hw_ctx_metrics{hwctx_handle, uuid, {}, {}});
+  }
+  catch(...) {
+    // dont log anything
+  }
+}
+
+void
+usage_metrics_logger::
+log_buffer_info_construct(device_id dev_id, size_t sz, const xrt_core::hwctx_handle* handle)
+{
+  struct bo_metrics* bo_met = nullptr;
+
+  auto map_it = m_dev_map.find(dev_id);
+  // dont log if device didn't match
+  if (map_it == m_dev_map.end())
+    return;
+  
+  auto& [device_id, dev_metrics] = *map_it;
+  if (!handle) {
+    // global bo
+    bo_met = &dev_metrics.global_bos_met;
```

Comment:
BO's that are created without using hw_context

Reply:
Got it.  I think it wouldn't hurt having two log_buffer_info_construct overloads, one for device and one for hwctx.  This assumes some refactoring for the shared pieces is possible.

Reply:
Soren if we implement overload function for global and hw ctx bo we will have 2 functions each for construct, destruct and sync bo. So using single function to avoid this. But I have refactored code to push common functionality into different function